dotCMS/src/main/java/com/dotcms/storage/FileSystemStoragePersistenceAPIImpl.java

```diff
@@ -0,0 +1,455 @@
+     * @param file       {@link File}   the actual file
+     * @param extraMeta  {@link Map} optional metadata, this could be null but depending on the implementation it would need some meta info.
+     * @return
+     */
+    @Override
+    public Future<Object> pushFileAsync(final String groupName, final String path,
+            final File file, final Map<String, Serializable> extraMeta) {
+        return DotConcurrentFactory.getInstance().getSubmitter(STORAGE_POOL).submit(
+                () -> this.pushFile(groupName, path, file, extraMeta)
+        );
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param groupName
+     * @param path       {@link String} path to upload the file
+     * @param writerDelegate     {@link ObjectWriterDelegate} stream to upload
+     * @param object     {@link Serializable} object to write into the storage
+     * @param extraMeta  {@link Map} optional metadata, this could be null but depending on the implementation it would need some meta info.
+     * @return
+     */
+    @Override
+    public Future<Object> pushObjectAsync(final String groupName, final String path,
+            final ObjectWriterDelegate writerDelegate, final Serializable object,
+            final Map<String, Serializable> extraMeta) {
+
+        return DotConcurrentFactory.getInstance().getSubmitter(STORAGE_POOL).submit(
+                () -> this.pushObject(groupName, path, writerDelegate, object, extraMeta)
+        );
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param groupName {@link String} group name
+     * @param path {@link String}
+     * @return
+     */
+    @Override
+    public File pullFile(final String groupName, final String path) {
+
+        if (!this.existsGroup(groupName)) {
+
+            throw new IllegalArgumentException(String.format(
+                    THE_BUCKET_NAME_S_DOES_NOT_HAVE_ANY_FILE_MAPPED,groupName));
+        }
+        final File clientFile;
+        final File bucketFile = this.groups.get(groupName.toLowerCase());
+        if (bucketFile.canRead()) {
+            final File destBucketFile = new File(bucketFile, path.toLowerCase());
+            if (destBucketFile.exists()) {
+                clientFile = destBucketFile;
+            } else {
+                throw new IllegalArgumentException("The group: " + destBucketFile + ", does not exists.");
+            }
+        } else {
+            throw new IllegalArgumentException("The bucket: " + groupName + " could not be read");
+        }
+
+        return clientFile;
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param groupName {@link String}  group name to pull
+     * @param path {@link String} path to pull the file
+     * @param readerDelegate {@link ObjectReaderDelegate} to reads the object
+     * @return
+     */
+    @Override
+    public Object pullObject(final String groupName, final String path,
+            final ObjectReaderDelegate readerDelegate) {
+
+        Object object;
```

Comment:
Codacy found an issue: [Avoid declaring a variable if it is unreferenced before a possible exit point.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6220272)